lib/CodeGen/MemoryAllocator.cpp

```diff
@@ -138,49 +154,573 @@ uint64_t MemoryAllocator::allocate(uint64_t size, Handle handle,
+  std::vector<uint64_t> buffSizeArray(buffNum);
+
+  // Array with the start/stop time (both inclusive) for each buffer.
+  std::vector<uint64_t> buffTimeStartArray(buffNum);
+  std::vector<uint64_t> buffTimeStopArray(buffNum);
+
+  // The maximum total required memory of all the live buffers reached during
+  // all allocation time steps. Note that this is the best size any allocation
+  // algorithm can hope for and is used to compute the allocation efficiency.
+  uint64_t liveSizeMax = 0;
+
+  // Array with the total required memory of all the live buffers for each
+  // allocation time step.
+  std::vector<uint64_t> liveBuffSizeArray(allocNum);
+
+  // Array with lists of IDs of all the live buffers for each allocation time
+  // step.
+  std::vector<std::list<uint64_t>> liveBuffIdListArray(allocNum);
+
+  // Gather information.
+  {
+    uint64_t allocIdx = 0;
+    uint64_t liveBuffSize = 0;
+    std::list<uint64_t> liveBuffIdList;
+    for (const auto &alloc : allocList) {
+
+      // Current buffer handle and mapped ID.
+      auto buffHandle = alloc.handle_;
+      auto buffId = handleToIdMap[buffHandle];
+
+      // Current buffer size. We only use the buffer size of an ALLOC request.
+      // For a FREE request we use the buffer size of the associated ALLOC.
+      // We round the requested buffer size using the alignment.
+      uint64_t buffSize;
+      if (alloc.alloc_) {
+        buffSize = getEffectiveSize(alloc.size_);
+      } else {
+        buffSize = buffSizeArray[buffId];
+      }
+
+      // Update buffer information.
+      if (alloc.alloc_) {
+        buffSizeArray[buffId] = buffSize;
+        buffTimeStartArray[buffId] = allocIdx;
+      } else {
+        buffTimeStopArray[buffId] = allocIdx;
+      }
+
+      // Update liveness information.
+      if (alloc.alloc_) {
+        liveBuffSize = liveBuffSize + buffSize;
+        liveBuffIdList.push_back(buffId);
+      } else {
+        liveBuffSize = liveBuffSize - buffSize;
+        auto it =
+            std::find(liveBuffIdList.begin(), liveBuffIdList.end(), buffId);
+        assert(it != liveBuffIdList.end() &&
+               "Buffer ID not found for removal!");
+        liveBuffIdList.erase(it);
+      }
+      liveSizeMax = std::max(liveSizeMax, liveBuffSize);
+      liveBuffSizeArray[allocIdx] = liveBuffSize;
+      liveBuffIdListArray[allocIdx] = liveBuffIdList;
+      allocIdx++;
+    }
+    assert(liveBuffSize == 0 &&
+           "Mismatch between total allocated and deallocated size!");
+    assert(liveBuffIdList.empty() &&
+           "Mismatch between total allocated and deallocated buffers!");
+  }
+
+  // If the theoretical required memory is larger than the available memory size
+  // then we return early.
+  if (memorySize_ && (liveSizeMax > memorySize_)) {
+    return MemoryAllocator::npos;
+  }
+
+  // ---------------------------------------------------------------------------
+  // Allocate all the buffers using all the available strategies.
```

Comment:
How does it affect the compilation time if you try to allocate using all available strategies?

Reply:
The function tries to use all the strategies but does not necessarily use them all. If the first strategy obtains the optimal allocation (with maximum efficiency) then the others are not used anymore. This is the case for most of the models (I think 70 out of the 79 models in the unit tests) the maximum efficiency is obtained with the first strategy.
Regarding compile-time the algorithm runs in a couple of ms.
e.g. The maximum it takes for the Model67 (resnet_152) which has lots of layers and for my VM it takes around 17 ms (the compile-time itself is in the order of 10's of seconds so this should be insignificant).